Implement retry mechanism to stabilise session - SessionNotCreatedException

**Is your feature request related to a problem? Please describe.**
Currently we have some amount of broken tests because browser can't be opened at the very beginning and step `Given I am on page with URL "URL"` is broken

**Describe the solution you'd like**
We would like to have retry mechanism for such cases.

**Additional context**
Error:
```
org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Possible causes are invalid address of the remote server or browser start-up failure. 
Host info: host: '5a27d612d5a9', ip: '172.17.0.2'
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:561)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:229)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:157)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:139)
	at org.vividus.selenium.RemoteWebDriverFactory.getRemoteWebDriver(RemoteWebDriverFactory.java:55)
	at org.vividus.selenium.AbstractWebDriverFactory.lambda$getRemoteWebDriver$3(AbstractWebDriverFactory.java:107)
	at org.vividus.selenium.AbstractWebDriverFactory.createWebDriver(AbstractWebDriverFactory.java:114)
	at org.vividus.selenium.AbstractWebDriverFactory.getRemoteWebDriver(AbstractWebDriverFactory.java:107)
	at org.vividus.selenium.VividusWebDriverFactory.createWebDriver(VividusWebDriverFactory.java:49)
	at org.vividus.selenium.AbstractVividusWebDriverFactory.create(AbstractVividusWebDriverFactory.java:57)
	at org.vividus.selenium.WebDriverProvider.get(WebDriverProvider.java:56)
	at org.vividus.ui.web.action.NavigateActions.getWebDriver(NavigateActions.java:116)
	at org.vividus.ui.web.action.NavigateActions.navigateTo(NavigateActions.java:45)
	at org.vividus.steps.ui.web.PageSteps.openPage(PageSteps.java:99)
	at jdk.internal.reflect.GeneratedMethodAccessor166.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.jbehave.core.steps.StepCreator$ParametrisedStep.perform(StepCreator.java:906)
	at org.jbehave.core.steps.StepCreator$ReportingAbstractStep.perform(StepCreator.java:649)
	at org.jbehave.core.embedder.PerformableTree$FineSoFar.run(PerformableTree.java:379)
	at org.jbehave.core.embedder.PerformableTree$FineSoFar.run(PerformableTree.java:397)
	at org.jbehave.core.embedder.PerformableTree$FineSoFar.run(PerformableTree.java:397)
	at org.jbehave.core.embedder.PerformableTree$PerformableSteps.perform(PerformableTree.java:1441)
	at org.jbehave.core.embedder.PerformableTree$AbstractPerformableScenario.lambda$performScenario$0(PerformableTree.java:1283)
	at org.jbehave.core.embedder.PerformableTree$PerformableEntity.performWithStopExecutionExceptionHandling(PerformableTree.java:965)
	at org.jbehave.core.embedder.PerformableTree$AbstractPerformableScenario.performScenario(PerformableTree.java:1276)
	at org.jbehave.core.embedder.PerformableTree$ExamplePerformableScenario.perform(PerformableTree.java:1353)
	at org.jbehave.core.embedder.PerformableTree$PerformableScenario.perform(PerformableTree.java:1220)
	at org.jbehave.core.embedder.PerformableTree$PerformableStory.lambda$perform$0(PerformableTree.java:1090)
	at org.jbehave.core.embedder.PerformableTree$PerformableEntity.performWithStopExecutionExceptionHandling(PerformableTree.java:965)
	at org.jbehave.core.embedder.PerformableTree$PerformableStory.perform(PerformableTree.java:1082)
	at org.jbehave.core.embedder.PerformableTree.performCancellable(PerformableTree.java:521)
	at org.jbehave.core.embedder.PerformableTree.perform(PerformableTree.java:488)
	at org.jbehave.core.embedder.StoryManager$EnqueuedStory.call(StoryManager.java:305)
	at org.jbehave.core.embedder.StoryManager$EnqueuedStory.call(StoryManager.java:278)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.openqa.selenium.TimeoutException: java.net.http.HttpConnectTimeoutException: HTTP connect timed out
Build info: version: '4.8.1', revision: '8ebccac989'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.0-1031-azure', java.version: '11.0.18'
Driver info: driver.version: RemoteWebDriver
	at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute(JdkHttpClient.java:294)
	at org.openqa.selenium.remote.tracing.TracedHttpClient.execute(TracedHttpClient.java:55)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:124)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:106)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:67)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:156)
	at org.openqa.selenium.remote.TracedCommandExecutor.execute(TracedCommandExecutor.java:51)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:543)
	... 38 more
Caused by: java.net.http.HttpConnectTimeoutException: HTTP connect timed out
	at java.net.http/jdk.internal.net.http.HttpClientImpl.send(Unknown Source)
	at java.net.http/jdk.internal.net.http.HttpClientFacade.send(Unknown Source)
	at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute(JdkHttpClient.java:292)
	... 45 more
Caused by: java.net.http.HttpConnectTimeoutException: HTTP connect timed out
	at java.net.http/jdk.internal.net.http.MultiExchange.toTimeoutException(Unknown Source)
	at java.net.http/jdk.internal.net.http.MultiExchange.getExceptionalCF(Unknown Source)
	at java.net.http/jdk.internal.net.http.MultiExchange.lambda$responseAsyncImpl$7(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
	at java.net.http/jdk.internal.net.http.Exchange.checkCancelled(Unknown Source)
	at java.net.http/jdk.internal.net.http.Exchange.cancel(Unknown Source)
	at java.net.http/jdk.internal.net.http.MultiExchange.cancel(Unknown Source)
	at java.net.http/jdk.internal.net.http.PlainHttpConnection$ConnectTimerEvent.handle(Unknown Source)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.purgeTimeoutsAndReturnNextDeadline(Unknown Source)
	at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.run(Unknown Source)
Caused by: java.net.ConnectException: HTTP connect timed out
	... 13 more
```

--- a/docs/modules/plugins/partials/selenium-properties.adoc
+++ b/docs/modules/plugins/partials/selenium-properties.adoc
@@ -1,3 +1,25 @@
+|[subs=+quotes]`*selenium.grid.host*`
+|hostname
+|`<empty>`
+|The host of https://www.selenium.dev/documentation/grid/[Selenium Grid] used to create a new session.
+
+|`selenium.grid.username`
+|username
+|`<empty>`
+|The username for authentication at https://www.selenium.dev/documentation/grid/[Selenium Grid] used to create a new session.
+
+|`selenium.grid.password`
+|password
+|`<empty>`
+|The password for authentication at https://www.selenium.dev/documentation/grid/[Selenium Grid] used to create a new session.
+
+|`selenium.grid.retry-session-creation-on-http-connect-timeout`
+a|`true` +
+`false`
+|`false`
+a|Whether to rety session creation at https://www.selenium.dev/documentation/grid/[Selenium Grid] in case of HTTP
+connect timeout. This flag may be useful when tests are executed via intermittent network connection.
+
 |`{session-scope-property-base}.session-scope`
 a|`story`
 
@@ -12,7 +34,7 @@ a|The test lifecycle of the {application-under-test}:
 An empty property value will lead to the error: "Application session scope is not set".
 
 |`ui.publish-source-on-failure`
-a|`true`
+a|`true` +
 `false`
 |`true`
 |Whether to publish the application source code on failure or not
@@ -22,21 +44,6 @@ a|`true` +
 `false`
 |`false`
 |If the value is set to `true`, an attempt to reset context will be performed when it becomes stale.
-
-|[subs=+quotes]`*selenium.grid.host*`
-|hostname
-|`<empty>`
-|Remote grid host to be used to create a new session
-
-|`selenium.grid.username`
-|username
-|`<empty>`
-|Remote grid username to be used to create a new session
-
-|`selenium.grid.password`
-|password
-|`<empty>`
-|Password to be used to create a new session
 
 |`ui.wait.timeout`
 |{durations-format-link} format

--- a/vividus-extension-selenium/src/main/java/org/vividus/selenium/RemoteWebDriverFactory.java
+++ b/vividus-extension-selenium/src/main/java/org/vividus/selenium/RemoteWebDriverFactory.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019-2022 the original author or authors.
+ * Copyright 2019-2023 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -17,9 +17,14 @@
 package org.vividus.selenium;
 
 import java.net.URL;
+import java.net.http.HttpConnectTimeoutException;
 
 import org.openqa.selenium.Capabilities;
+import org.openqa.selenium.SessionNotCreatedException;
+import org.openqa.selenium.TimeoutException;
 import org.openqa.selenium.remote.RemoteWebDriver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.vividus.selenium.manager.GenericWebDriverManager;
 
 import io.appium.java_client.android.AndroidDriver;
@@ -27,20 +32,48 @@ import io.appium.java_client.ios.IOSDriver;
 
 public class RemoteWebDriverFactory implements IRemoteWebDriverFactory
 {
-    private final boolean useW3C;
+    private static final Logger LOGGER = LoggerFactory.getLogger(RemoteWebDriverFactory.class);
 
-    public RemoteWebDriverFactory(boolean useW3C)
+    private final boolean useW3C;
+    private final boolean retrySessionCreationOnHttpConnectTimeout;
+
+    public RemoteWebDriverFactory(boolean useW3C, boolean retrySessionCreationOnHttpConnectTimeout)
     {
         this.useW3C = useW3C;
+        this.retrySessionCreationOnHttpConnectTimeout = retrySessionCreationOnHttpConnectTimeout;
     }
 
     @Override
     public RemoteWebDriver getRemoteWebDriver(URL url, Capabilities capabilities)
     {
-        /* Selenium 4 declares that it only supports W3C and nevertheless still writes  JWP's "desiredCapabilities" into
-        "createSession" JSON. Appium Java client eliminates that: https://github.com/appium/java-client/pull/1537.
-        But still some clouds (e.g. SmartBear CrossBrowserTesting) are not prepared for W3c, so we should avoid
-        using Appium drivers to create sessions for web tests. */
+        try
+        {
+            return createRemoteWebDriver(url, capabilities);
+        }
+        catch (SessionNotCreatedException e)
+        {
+            if (retrySessionCreationOnHttpConnectTimeout && isHttpConnectTimeout(e))
+            {
+                LOGGER.warn("Failed to create a new session due to HTTP connect timout", e);
+                LOGGER.warn("Retrying to create a new session");
+                return createRemoteWebDriver(url, capabilities);
+            }
+            throw e;
+        }
+    }
+
+    private static boolean isHttpConnectTimeout(SessionNotCreatedException e)
+    {
+        Throwable cause = e.getCause();
+        return cause instanceof TimeoutException && cause.getCause() instanceof HttpConnectTimeoutException;
+    }
+
+    private RemoteWebDriver createRemoteWebDriver(URL url, Capabilities capabilities)
+    {
+        /* Selenium 4 declares that it only supports W3C and nevertheless still writes  JWP's "desiredCapabilities"
+        into "createSession" JSON. Appium Java client eliminates that:https://github.com/appium/java-client/pull/1537.
+        But still some clouds (e.g. SmartBear CrossBrowserTesting) are not prepared for W3c, so we should avoid using
+        Appium drivers to create sessions for web tests. */
         if (useW3C)
         {
             if (GenericWebDriverManager.isIOS(capabilities) || GenericWebDriverManager.isTvOS(capabilities))

--- a/vividus-extension-selenium/src/main/resources/properties/defaults/default.properties
+++ b/vividus-extension-selenium/src/main/resources/properties/defaults/default.properties
@@ -29,6 +29,7 @@ proxy.publish-har-on-failure=false
 ui.screenshot.debug-directory=
 
 selenium.grid.use-w3c=true
+selenium.grid.retry-session-creation-on-http-connect-timeout=false
 
 ui.visual.ignored-elements=
 ui.visual.ignored-areas=

--- a/vividus-extension-selenium/src/main/resources/vividus-extension/spring.xml
+++ b/vividus-extension-selenium/src/main/resources/vividus-extension/spring.xml
@@ -20,6 +20,7 @@
 
     <bean id="remoteWebDriverFactory" class="org.vividus.selenium.RemoteWebDriverFactory">
         <constructor-arg value="${selenium.grid.use-w3c}" />
+        <constructor-arg value="${selenium.grid.retry-session-creation-on-http-connect-timeout}" />
     </bean>
 
     <bean id="webDriverProvider" class="org.vividus.selenium.WebDriverProvider"

--- a/vividus-extension-selenium/src/test/java/org/vividus/selenium/RemoteWebDriverFactoryTests.java
+++ b/vividus-extension-selenium/src/test/java/org/vividus/selenium/RemoteWebDriverFactoryTests.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019-2022 the original author or authors.
+ * Copyright 2019-2023 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -16,92 +16,156 @@
 
 package org.vividus.selenium;
 
+import static com.github.valfirst.slf4jtest.LoggingEvent.warn;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.is;
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mockConstruction;
-import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.when;
 
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.net.URL;
+import java.net.http.HttpConnectTimeoutException;
+import java.net.http.HttpTimeoutException;
 import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import com.github.valfirst.slf4jtest.TestLogger;
+import com.github.valfirst.slf4jtest.TestLoggerFactory;
+import com.github.valfirst.slf4jtest.TestLoggerFactoryExtension;
 
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 import org.mockito.Mock;
-import org.mockito.MockedConstruction;
-import org.mockito.MockedStatic;
 import org.mockito.junit.jupiter.MockitoExtension;
-import org.openqa.selenium.Capabilities;
+import org.openqa.selenium.Platform;
+import org.openqa.selenium.SessionNotCreatedException;
+import org.openqa.selenium.TimeoutException;
+import org.openqa.selenium.remote.CapabilityType;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.HttpCommandExecutor;
 import org.openqa.selenium.remote.RemoteWebDriver;
-import org.vividus.selenium.manager.GenericWebDriverManager;
+import org.openqa.selenium.remote.Response;
 
 import io.appium.java_client.android.AndroidDriver;
 import io.appium.java_client.ios.IOSDriver;
+import io.appium.java_client.remote.MobilePlatform;
 
-@ExtendWith(MockitoExtension.class)
+@ExtendWith({ MockitoExtension.class, TestLoggerFactoryExtension.class })
 class RemoteWebDriverFactoryTests
 {
+    private static final URI GRID_URI = URI.create("http://sel.grid");
+
+    private static final String COULD_NOT_START_A_NEW_SESSION = "Could not start a new session";
+    private static final String HTTP_CONNECT_TIMED_OUT = "HTTP connect timed out";
+
     @Mock private URL url;
-    @Mock private Capabilities capabilities;
 
-    @SuppressWarnings("rawtypes")
-    @Test
-    void shouldCreateDriverForIOS()
+    private final TestLogger logger = TestLoggerFactory.getTestLogger(RemoteWebDriverFactory.class);
+
+    static Stream<Arguments> platforms()
     {
-        try (MockedStatic<GenericWebDriverManager> manager = mockStatic(GenericWebDriverManager.class))
-        {
-            manager.when(() -> GenericWebDriverManager.isIOS(capabilities)).thenReturn(true);
-            testWebDriverCreation(true, IOSDriver.class);
-        }
+        return Stream.of(
+                arguments(Platform.IOS, true, IOSDriver.class),
+                arguments(Platform.IOS, false, RemoteWebDriver.class),
+                arguments(MobilePlatform.TVOS, true, IOSDriver.class),
+                arguments(MobilePlatform.TVOS, false, RemoteWebDriver.class),
+                arguments(Platform.ANDROID, true, AndroidDriver.class),
+                arguments(Platform.ANDROID, false, RemoteWebDriver.class),
+                arguments(Platform.WINDOWS, true, RemoteWebDriver.class),
+                arguments(Platform.WINDOWS, false, RemoteWebDriver.class)
+        );
     }
 
-    @Test
-    void shouldCreateRemoteWebDriverWhenNonW3cProtocolIsUsed()
+    @MethodSource("platforms")
+    @ParameterizedTest
+    void shouldCreateDriver(Object platform, boolean useW3C, Class<?> webDriveClass)
     {
-        testWebDriverCreation(false, RemoteWebDriver.class);
-    }
-
-    @SuppressWarnings("rawtypes")
-    @Test
-    void shouldCreateDriverForTvOS()
-    {
-        try (MockedStatic<GenericWebDriverManager> manager = mockStatic(GenericWebDriverManager.class))
+        var capabilities = new DesiredCapabilities();
+        capabilities.setCapability(CapabilityType.PLATFORM_NAME, platform);
+        try (var driver = mockConstruction(webDriveClass,
+                (mock, context) -> assertEquals(context.arguments(), List.of(url, capabilities))))
         {
-            manager.when(() -> GenericWebDriverManager.isIOS(capabilities)).thenReturn(false);
-            manager.when(() -> GenericWebDriverManager.isTvOS(capabilities)).thenReturn(true);
-            testWebDriverCreation(true, IOSDriver.class);
-        }
-    }
-
-    @SuppressWarnings("rawtypes")
-    @Test
-    void shouldCreateDriverForAndroid()
-    {
-        try (MockedStatic<GenericWebDriverManager> manager = mockStatic(GenericWebDriverManager.class))
-        {
-            manager.when(() -> GenericWebDriverManager.isIOS(capabilities)).thenReturn(false);
-            manager.when(() -> GenericWebDriverManager.isTvOS(capabilities)).thenReturn(false);
-            manager.when(() -> GenericWebDriverManager.isAndroid(capabilities)).thenReturn(true);
-            testWebDriverCreation(true, AndroidDriver.class);
-        }
-    }
-
-    @Test
-    void shouldCreateRemoteWebDriver()
-    {
-        try (MockedStatic<GenericWebDriverManager> manager = mockStatic(GenericWebDriverManager.class))
-        {
-            manager.when(() -> GenericWebDriverManager.isIOS(capabilities)).thenReturn(false);
-            manager.when(() -> GenericWebDriverManager.isTvOS(capabilities)).thenReturn(false);
-            manager.when(() -> GenericWebDriverManager.isAndroid(capabilities)).thenReturn(false);
-            testWebDriverCreation(true, RemoteWebDriver.class);
-        }
-    }
-
-    private <T> void testWebDriverCreation(boolean useW3C, Class<T> webDriveClass)
-    {
-        try (MockedConstruction<T> driver = mockConstruction(webDriveClass))
-        {
-            var actualDriver = new RemoteWebDriverFactory(useW3C).getRemoteWebDriver(url, capabilities);
+            var actualDriver = new RemoteWebDriverFactory(useW3C, false).getRemoteWebDriver(url, capabilities);
             assertEquals(driver.constructed(), List.of(actualDriver));
+            assertThat(logger.getLoggingEvents(), is(empty()));
+        }
+    }
+
+    @Test
+    void shouldRetrySessionCreationOnHttpConnectTimeoutSuccessfully() throws URISyntaxException
+    {
+        var capabilities = Map.of(CapabilityType.PLATFORM_NAME, Platform.LINUX.name());
+        var desiredCapabilities = new DesiredCapabilities(capabilities);
+        var sessionId = "mocked-session-id";
+        var exception = new SessionNotCreatedException(COULD_NOT_START_A_NEW_SESSION,
+                new TimeoutException(HTTP_CONNECT_TIMED_OUT, new HttpConnectTimeoutException(HTTP_CONNECT_TIMED_OUT))
+        );
+        try (var ignored = mockConstruction(HttpCommandExecutor.class, (mock, context) -> {
+            if (context.getCount() == 1)
+            {
+                when(mock.execute(any())).thenThrow(exception);
+            }
+            else
+            {
+                var response = new Response();
+                response.setValue(capabilities);
+                response.setSessionId(sessionId);
+                when(mock.execute(any())).thenReturn(response);
+            }
+        }))
+        {
+            when(url.toURI()).thenReturn(GRID_URI);
+            var actualDriver = new RemoteWebDriverFactory(true, true).getRemoteWebDriver(url, desiredCapabilities);
+            assertEquals(sessionId, actualDriver.getSessionId().toString());
+            assertThat(logger.getLoggingEvents(), is(List.of(
+                    warn(exception, "Failed to create a new session due to HTTP connect timout"),
+                    warn("Retrying to create a new session")
+            )));
+        }
+    }
+
+    static Stream<Arguments> exceptions()
+    {
+        return Stream.of(
+                arguments(new SessionNotCreatedException(COULD_NOT_START_A_NEW_SESSION,
+                        new TimeoutException("HttpTimeoutException: HTTP read timed out",
+                                new HttpTimeoutException("HTTP read timed out")
+                        )), true),
+                arguments(new SessionNotCreatedException(COULD_NOT_START_A_NEW_SESSION, new IOException("IOException")),
+                        true),
+                arguments(new SessionNotCreatedException(COULD_NOT_START_A_NEW_SESSION,
+                        new TimeoutException(HTTP_CONNECT_TIMED_OUT,
+                                new HttpConnectTimeoutException(HTTP_CONNECT_TIMED_OUT))), false)
+        );
+    }
+
+    @MethodSource("exceptions")
+    @ParameterizedTest
+    void shouldRetrySessionCreationOnHttpConnectTimeoutUnsuccessfully(SessionNotCreatedException exception,
+            boolean retrySessionCreationOnHttpConnectTimeout) throws URISyntaxException
+    {
+        var capabilities = Map.of(CapabilityType.PLATFORM_NAME, Platform.LINUX.name());
+        var desiredCapabilities = new DesiredCapabilities(capabilities);
+        try (var ignored = mockConstruction(HttpCommandExecutor.class,
+                (mock, context) -> when(mock.execute(any())).thenThrow(exception)))
+        {
+            when(url.toURI()).thenReturn(GRID_URI);
+            var remoteWebDriverFactory = new RemoteWebDriverFactory(true, retrySessionCreationOnHttpConnectTimeout);
+            var actualException = assertThrows(SessionNotCreatedException.class,
+                    () -> remoteWebDriverFactory.getRemoteWebDriver(url, desiredCapabilities));
+            assertEquals(exception, actualException);
+            assertThat(logger.getLoggingEvents(), is(empty()));
         }
     }
 }

